internal/statistics.go

```diff
@@ -0,0 +1,231 @@
+	return &stats
+}
+
+// start registers all client statistics and schedules periodic collection of stats.
+func (s *statistics) start() {
+	if !s.enabled {
+		return
+	}
+	s.registerMetrics()
+	period := s.client.properties.GetDuration(property.StatisticsPeriodSeconds)
+	if period <= 0 {
+		// This will return the default value since we have a non-positive value now
+		defaultPeriod := s.client.properties.GetPositiveDurationOrDef(property.StatisticsPeriodSeconds)
+		log.Println("Provided client statistics ", property.StatisticsPeriodSeconds.Name(),
+			" cannot be less than or equal to 0.",
+			" You provided ", period, " as the configuration. Client will use the default value ", defaultPeriod,
+			" instead.")
+		period = defaultPeriod
+	}
+	go s.schedulePeriodicStatisticsSendTask(period)
+}
+
+func (s *statistics) schedulePeriodicStatisticsSendTask(period time.Duration) {
+	s.sendStatistics()
+	ticker := time.NewTicker(period)
+	for {
+		select {
+		case <-s.cancel:
+			return
+		case <-ticker.C:
+			s.sendStatistics()
+		}
+	}
+}
+
+func (s *statistics) getOwnerConnection() *Connection {
+	connection := s.client.ConnectionManager.getOwnerConnection()
+	if connection == nil {
+		return nil
+	}
+	ownerConnectionAddress := s.client.ClusterService.getOwnerConnectionAddress()
+	serverVersion := connection.connectedServerVersion
+	if serverVersion < sinceVersion {
+		// do not print too many logs if connected to an old version server
+		cachedOwnerAddress, ok := s.ownerAddress.Load().(*proto.Address)
+		if !ok || (ownerConnectionAddress != cachedOwnerAddress) {
+			log.Println("Client statistics cannot be sent to server ", ownerConnectionAddress,
+				" since connected server version is less than the minimum supported server version ",
+				sinceVersionString)
+		}
+
+		// cache the last connected server address for decreasing the log prints
+		s.ownerAddress.Store(ownerConnectionAddress)
+		return nil
+	}
+	return connection
+}
+
+var emptyStatValueFunc = func() string {
+	return emptyStatValue
+}
+
+func (s *statistics) sendStatistics() {
+	ownerConnection := s.getOwnerConnection()
+	if ownerConnection == nil {
+		log.Println("Cannot send client statistics to the server. No owner connection")
+		return
+	}
+
+	stats := &bytes.Buffer{}
+
+	s.fillMetrics(stats, ownerConnection)
+	s.sendStatsToOwner(stats)
+}
+
+func (s *statistics) sendStatsToOwner(stats *bytes.Buffer) {
+	request := proto.ClientStatisticsEncodeRequest(stats.String())
+	ownerConnection := s.getOwnerConnection()
+	_, err := s.client.InvocationService.invokeOnConnection(request, ownerConnection).Result()
+	if err != nil {
+		log.Println("Could not send the statistics, ", err)
+	}
+}
+
+func (s *statistics) registerMetrics() {
+	var m runtime.MemStats
+	runtime.ReadMemStats(&m)
+	s.registerGauge("os.committedVirtualMemorySize",
+		func() string {
+			return strconv.Itoa(int(m.Sys))
+		})
+	s.registerGauge("os.freePhysicalMemorySize", emptyStatValueFunc)
```

Comment:
a lot of statistics is registered as empty stat, why? does "go" have no way to obtain os statistics? If it is empty why are you registering? Better not to register at all.

Reply:
I added some more but I couldnt find some of the statistics.

Reply:
Do not add if no such statistics, this is misleading as if it were implemented. If the statistics do not exist do not register empty stat. I believe the there should be no registration of empty stat.

Reply:
I think in Non java clients, missing statistics are added as empty stat. I am not sure what the exact reason for this is though, @asimarslan ?